src/backend/distributed/commands/create_citus_local_table.c

```diff
@@ -0,0 +1,620 @@
+/*-------------------------------------------------------------------------
+ *
+ * create_citus_local_table.c
+ *
+ * This file contains functions to create citus local tables.
+ *
+ * Copyright (c) Citus Data, Inc.
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#include "postgres.h"
+
+#include "access/genam.h"
+#include "access/htup_details.h"
+#include "catalog/pg_constraint.h"
+#include "distributed/citus_local_table_utils.h"
+#include "distributed/citus_ruleutils.h"
+#include "distributed/colocation_utils.h"
+#include "distributed/commands.h"
+#include "distributed/commands/utility_hook.h"
+#include "distributed/listutils.h"
+#include "distributed/master_metadata_utility.h"
+#include "distributed/master_protocol.h"
+#include "distributed/metadata_sync.h"
+#include "distributed/multi_partitioning_utils.h"
+#include "distributed/reference_table_utils.h"
+#include "distributed/worker_protocol.h"
+#include "utils/builtins.h"
+#include "utils/fmgroids.h"
+#include "utils/lsyscache.h"
+#include "utils/syscache.h"
+
+static void ErrorIfUnsupportedCreateCitusLocalTable(Oid relationId);
+static void CreateCitusLocalTableShard(Oid relationId, uint64 shardId);
+static void RenameRelationToShardRelation(Oid shellRelationId, uint64 shardId);
+static void RenameShardRelationConstraints(Oid shardRelationId, uint64 shardId);
+static List * GetConstraintNameList(Oid relationId);
+static void RenameForeignConstraintsReferencingToShard(Oid shardRelationId,
+													   uint64 shardId);
+static void RenameShardRelationIndexes(Oid shardRelationId, uint64 shardId);
+static List * GetExplicitIndexNameList(Oid relationId);
+static void CreateCitusLocalTable(Oid shellRelationId);
+static void CreateShellTableForCitusLocalTable(List *shellTableDDLEvents);
+static void InsertMetadataForCitusLocalTable(Oid citusLocalTableId, uint64 shardId);
+
+PG_FUNCTION_INFO_V1(create_citus_local_table);
+
+/*
+ * create_citus_local_table creates a citus table from the table with relationId.
+ * The created table would have the following properties:
+ *  - it will have only one shard,
+ *  - its distribution method will be DISTRIBUTE_BY_NONE,
+ *  - its replication model will be ReplicationModel,
+ *  - its replication factor will be set to 1.
+ * Similar to reference tables, it has only 1 placement. In addition to that, that
+ * single placement is only allowed to be on the coordinator.
+ */
+Datum
+create_citus_local_table(PG_FUNCTION_ARGS)
+{
+	CheckCitusVersion(ERROR);
+
+	Oid relationId = PG_GETARG_OID(0);
+
+	/*
+	 * Lock target relation with an exclusive lock - there's no way to make
+	 * sense of this table until we've committed, and we don't want multiple
+	 * backends manipulating this relation.
+	 */
+	Relation relation = try_relation_open(relationId, ExclusiveLock);
```

Comment:
It is probably to give a "slightly" better error message claiming that we couldn't create the table too. But, on the other side, it would also be a good option to leave those kind of native checks to postgres to be more compliant in terms of the error messages.

Reply:
Yeah, I feel like the error we give is not better than what postgres gives, but it is not related to this PR.